Rewrite concat

natebosch opened the pull request:
Uses sync forwarders for dart-lang/tools#1754

The simple implementation leaves out a lot of edge cases around
single-subscription vs broadcast streams and how they combine. Rewrite
to handle pause, resume, and cancel manually.

- Loop over stream types in tests
- Add test for onDone and canceling and relistening
- Handle mix of stream types by choosign whether to cancel or pause as
  appropriate.

natebosch committed:
Rewrite concat

Uses sync forwarders for #24

The simple implementation leaves out a lot of edge cases around
single-subscription vs broadcast streams and how they combine. Rewrite
to handle pause, resume, and cancel manually.

- Loop over stream types in tests
- Add test for onDone and canceling and relistening
- Handle mix of stream types by choosign whether to cancel or pause as
  appropriate.

jakemac53 commented on lib/src/concat.dart:
```diff
@@ -17,11 +20,66 @@ class _Concat<T> implements StreamTransformer<T, T> {
+    currentDoneHandler = onFirstDone;
+
+    controller.onListen = () {
+      if (firstDone &&
```
I am a bit confused by a lot of the logic here, why do we only care if `_next` is a broadcast stream, shouldn't most of these be checking for `currentStream.isBroadcast`?

natebosch replied:
The code paths around re-listening should only get hit if `first` is a broadcast stream, otherwise the controller would prevent us from ever getting here.

Our first listener can't hit this since `firstDone` won't ever be true. Subsequent listeners can only hit if `first` was a broadcast stream. We're not still listening to `first` since `firstDone` is true, which means we are (were) listening to `_next` so it's the only stream that `subscription` could be on.

natebosch replied:
Using `.asBroadcastStream()` allowed me to drop this block

jakemac53 commented on test/concat_test.dart:
```diff
@@ -8,40 +8,121 @@ import 'package:test/test.dart';
+          first = streamTypes[firstType]()
+            ..onCancel = () {
+              firstCanceled = true;
+            };
```
should we be tracking `onListen` for this as well?

natebosch replied:
I didn't think it was critical since the test that values come through prove we've listened when we should.

jakemac53 commented on test/concat_test.dart:
```diff
@@ -8,40 +8,121 @@ import 'package:test/test.dart';
+          expect(secondListened, true);
+        });
+
+        test('cancels single-subscription first stream on cancel', () async {
```
Should this be inside an `if (firstType == 'single-subscription')` block (the test name indicates that at least).

jakemac53 commented on test/concat_test.dart:
```diff
@@ -8,40 +8,121 @@ import 'package:test/test.dart';
+          expect(firstCanceled, true);
+        });
+
+        if (firstType == 'single-subscription') {
```
I would expect this to be `secondType == 'single-subscription'` based on the test name?

natebosch replied:
I might need to rename it. If the first stream is broadcast then a single-subscription second stream never gets canceled - only paused.

jakemac53 commented on test/concat_test.dart:
```diff
@@ -8,40 +8,121 @@ import 'package:test/test.dart';
+          expect(isDone, true);
+        });
+
+        if (firstType == 'broadcast') {
```
it might be worth adding a `group` here with the name `firstType == 'broadcast'` or something

natebosch replied:
The current group names both types of streams so I don't think it would help the output. IMO the group would add a level of indent without much benefit.

jakemac53 commented on test/concat_test.dart:
```diff
@@ -8,40 +8,121 @@ import 'package:test/test.dart';
+          first..add(1)..add(2);
+          await first.close();
+          await new Future(() {});
+          second..add(3)..add(4);
```
What if these are added before `first` closes? It seems like in the broadcast stream case you wouldn't catch any of those values?

natebosch replied:
Correct. A lot of behavior in this transformer don't make much sense for broadcast streams, especially so for mixing them, but I thought it'd be best to handle it as well as possible. I can expand the doc comment to call some of this stuff out.

jakemac53 replied:
Technically if it is a broadcast stream it could start collecting things right away, and just not report them until the first stream is done right? I don't know if that is what users would expect or not.

natebosch replied:
Expanded the doc comment a bit.

natebosch replied:
Somehow I missed the reply earlier.

I don't think it would be best to buffer - it's a potential memory leak. If we have to pick one of 2 bad behaviors I think missing earlier events is the less bad choice.

It's also possible to get that behavior manually by forcing back to single-subscription. Whereas it's not possible to get back to the other behavior if we were to buffer.

```dart
Stream<T> singleSubscriptionEarlyListen<T>(Stream<T> values) {
  var controller = new StreamController<T>();
  values.listen(controller.add, onError: controller.adError, onDone: controller.close);
  return controller.stream;
}
//...
someStream.transform(concat(singleSubscriptionEarlyListen(someBroadcast)));
```

natebosch reviewed (commented):
One thing I'm not sure of - If a single subscription stream is `concat`ed with a broadcast stream, then a cancel will cause it to `pause` - buffer values until it get's listened again.

Maybe instead I should replace the listener with something that ignores values instead?

natebosch commented:
Come to think of it - what I did with `merge` was to use `asBroadcastStream()` for any single-subscription stuff that get's merged in... maybe that's what I should do here.

natebosch committed:
Force _next to broadcast when first is broadcast

Drop some now unnecessary extra logic around handling a
single-subscription stream following a broadcast stream.

natebosch committed:
Document behavior of broadcast stream as second stream, expand test for pause and re-listen to show that events are dropped

natebosch committed:
Rename tests for clarity

natebosch committed:
Reorder so all conditional tests are at the end

natebosch committed:
Add tests for pause/resume

natebosch commented:
@jakemac53 any other thoughts?
